crates/ruff_server/src/session/index.rs

```diff
@@ -347,16 +347,10 @@ impl Index {
             anyhow::bail!("Tried to close unavailable document `{key}`");
         };
 
-        let Some(controller) = self.documents.remove(&url) else {
+        if self.documents.remove(&url).is_none() {
             anyhow::bail!("tried to close document that didn't exist at {}", url)
         };
-        if let Some(notebook) = controller.as_notebook() {
-            for url in notebook.urls() {
-                self.notebook_cells.remove(url).ok_or_else(|| {
-                    anyhow!("tried to de-register notebook cell with URL {url} that didn't exist")
-                })?;
-            }
-        }
+        // notebook cells are closed via textDocument/didClose - we don't close them here.
```

Comment:
Does that mean that VS code sends an explicit `didClose` request before (or after) for the cells?

Reply:
Yes, it always will send `textDocument/didClose` either before or after.

* I noticed on regular `file://` urls it sends `textDocument/didClose` per cell _before_ `notebookDocument/didClose`:

```log
2024-06-14 16:46:48.544 [info] [Trace - 4:46:48 PM] Sending notification 'textDocument/didClose'.
2024-06-14 16:46:48.544 [info] Params: {
    "textDocument": {
        "uri": "vscode-notebook-cell:/e%3A/test/test.ipynb#W0sZmlsZQ%3D%3D"
    }
}


2024-06-14 16:46:48.544 [info] [Trace - 4:46:48 PM] Sending notification 'textDocument/didClose'.
2024-06-14 16:46:48.544 [info] Params: {
    "textDocument": {
        "uri": "vscode-notebook-cell:/e%3A/test/test.ipynb#W1sZmlsZQ%3D%3D"
    }
}


2024-06-14 16:46:48.544 [info] [Trace - 4:46:48 PM] Sending notification 'notebookDocument/didClose'.
2024-06-14 16:46:48.544 [info] Params: {
    "notebookDocument": {
        "uri": "file:///e%3A/test/test.ipynb"
    },
    "cellTextDocuments": [
        {
            "uri": "vscode-notebook-cell:/e%3A/test/test.ipynb#W0sZmlsZQ%3D%3D"
        },
        {
            "uri": "vscode-notebook-cell:/e%3A/test/test.ipynb#W1sZmlsZQ%3D%3D"
        }
    ]
}


2024-06-14 16:46:48.545 [info] [Trace - 4:46:48 PM] Received notification 'textDocument/publishDiagnostics'.
2024-06-14 16:46:48.545 [info] Params: {
    "diagnostics": [],
    "uri": "vscode-notebook-cell:/e%3A/test/test.ipynb#W0sZmlsZQ%3D%3D",
    "version": 2
}


2024-06-14 16:46:48.545 [info] [Trace - 4:46:48 PM] Received notification 'textDocument/publishDiagnostics'.
2024-06-14 16:46:48.545 [info] Params: {
    "diagnostics": [],
    "uri": "vscode-notebook-cell:/e%3A/test/test.ipynb#W1sZmlsZQ%3D%3D",
    "version": 2
}
```

* For "untitled:Untitled-1.ipynb" and "untitled:/Interactive-2.interactive" urls it will send `textDocument/didClose` per cell _after_ `notebookDocument/didClose`:

```log
2024-06-14 16:58:23.344 [info] [Trace - 4:58:23 PM] Sending notification 'notebookDocument/didClose'.
2024-06-14 16:58:23.344 [info] Params: {
    "notebookDocument": {
        "uri": "untitled:Untitled-1.ipynb?jupyter-notebook"
    },
    "cellTextDocuments": [
        {
            "uri": "vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D"
        }
    ]
}


2024-06-14 16:58:23.364 [info] [Trace - 4:58:23 PM] Sending notification 'textDocument/didClose'.
2024-06-14 16:58:23.364 [info] Params: {
    "textDocument": {
        "uri": "vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D"
    }
}


2024-06-14 16:58:23.365 [info] 1173.763966200s ERROR ruff:main ruff_server::server::api: An error occurred while running textDocument/didClose: Unable to take snapshot for document with URL vscode-notebook-cell:Untitled-1.ipynb?jupyter-notebook#W0sdW50aXRsZWQ%3D

2024-06-14 16:58:23.365 [info] [Trace - 4:58:23 PM] Received notification 'window/showMessage'.
2024-06-14 16:58:23.365 [info] Params: {
    "message": "Ruff encountered a problem. Check the logs for more details.",
    "type": 1
}
```